File: examples/employee-directory-laravel/tests/Feature/AuthAndOrgTest.php
<?php

namespace Tests\Feature;

use App\Models\Organization;
use App\Models\User;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Tests\TestCase;

class AuthAndOrgTest extends TestCase
{
    use RefreshDatabase;

    private function register(string $email = 'anna@example.com'): User
    {
        $this->post('/register', [
            'email' => $email, 'password' => 'sikkerhed123', 'name' => 'Anna',
        ]);

        return User::where('email', $email)->firstOrFail();
    }

    public function test_register_signs_in_and_lands_home(): void
    {
        $response = $this->post('/register', [
            'email' => 'ny@example.com', 'password' => 'sikkerhed123',
        ]);
        $response->assertRedirect('/');
        $this->assertAuthenticated();
    }

    public function test_login_rejects_wrong_password(): void
    {
        $this->register();
        auth()->logout();
        $response = $this->post('/login', [
            'email' => 'anna@example.com', 'password' => 'forkert',
        ]);
        $response->assertSessionHasErrors('email');
    }

    public function test_creating_an_organization_scaffolds_checklist_and_owner_role(): void
    {
        $user = $this->register();
        $this->post('/orgs', ['name' => 'Personalehuset ApS']);
        $organization = Organization::where('name', 'Personalehuset ApS')->firstOrFail();

        $this->assertSame('owner', $organization->roleOf($user));
        $this->assertSame(3, $organization->onboardingTasks()->count());
    }

    public function test_last_owner_cannot_be_demoted_or_removed(): void
    {
        $this->register();
        $this->post('/orgs', ['name' => 'Solo']);
        $organization = Organization::where('name', 'Solo')->firstOrFail();
        $membership = $organization->memberships()->first();

        $this->patch("/orgs/{$organization->slug}/members/{$membership->id}", ['role' => 'member']);
        $this->assertSame('owner', $membership->fresh()->role);

        $this->delete("/orgs/{$organization->slug}/members/{$membership->id}");
        $this->assertSame('active', $membership->fresh()->status);
    }

    public function test_non_members_get_403(): void
    {
        $this->register();
        $this->post('/orgs', ['name' => 'Privat']);
        $organization = Organization::where('name', 'Privat')->firstOrFail();

        auth()->logout();
        $this->register('fremmed@example.com');
        $this->get("/orgs/{$organization->slug}")->assertForbidden();
        $this->get("/orgs/{$organization->slug}/employees")->assertForbidden();
    }
}

File: examples/crm-rails/test/integration/auth_and_org_test.rb
require "test_helper"

class AuthAndOrgTest < ActionDispatch::IntegrationTest
  def register!(email = "anna@example.com")
    post "/register", params: { email: email, password: "sikkerhed123", full_name: "Anna" }
    User.find_by!(email: email)
  end

  test "register signs in and lands home" do
    register!
    assert_redirected_to "/"
  end

  test "login rejects wrong password" do
    register!
    delete "/logout"
    post "/login", params: { email: "anna@example.com", password: "forkert" }
    assert_response :unprocessable_entity
    assert_match "Wrong email or password", response.body
  end

  test "creating an organization scaffolds pipeline stages and owner role" do
    register!
    post "/organizations", params: { name: "Kystvej Consulting" }
    organization = Organization.find_by!(name: "Kystvej Consulting")
    assert_equal "owner", organization.role_of(User.find_by!(email: "anna@example.com"))
    assert_equal %w[Lead Qualified Proposal Won],
                 organization.pipelines.first.stages.map(&:name)
  end

  test "the last owner cannot be demoted or removed" do
    register!
    post "/organizations", params: { name: "Solo" }
    organization = Organization.find_by!(name: "Solo")
    membership = organization.memberships.first

    patch "/orgs/#{organization.slug}/members/#{membership.id}", params: { role: "member" }
    assert_equal "owner", membership.reload.role

    delete "/orgs/#{organization.slug}/members/#{membership.id}"
    assert_equal "active", membership.reload.status
  end

  test "non-members get 403 on organization pages" do
    register!
    post "/organizations", params: { name: "Privat" }
    organization = Organization.find_by!(name: "Privat")

    delete "/logout"
    register!("fremmed@example.com")
    get "/orgs/#{organization.slug}"
    assert_response :forbidden
    get "/orgs/#{organization.slug}/contacts"
    assert_response :forbidden
  end
end
